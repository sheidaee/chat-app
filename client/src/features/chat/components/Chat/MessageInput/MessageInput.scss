@import "../../../../../sass/abstracts/theme";

/* message input */
.messageInput {
  @include themify($themes) {
    background: themed('messageInputBgColor');
    color: themed('messageInputColor');
  }
	height: 10vh;
	max-height: 6.5rem;

  .messageForm {
    height: 100%;
    width: 100%;
    display: flex;
    justify-content: space-between;

    .formControl {
      padding-top: 2.4rem;
      padding-bottom: 2.4rem;
      resize: none;
      padding-left: 1.5rem;
      box-sizing: border-box;
      width: 80%;
      height: 100%;
      border: none;

      // &::-webkit-scrollbar-track {
      //   -webkit-box-shadow: inset 0 0 0.6rem rgba(0, 0, 0, 0.3);
      //   background-color: #cac8ee;
      // }

      // &::-webkit-scrollbar {
      //   width: 0.5rem;
      //   background-color: green;
      // }

      // &::-webkit-scrollbar-thumb {
      //   background-color: #484d79;
      // }

      &:focus {
        outline: none;
        box-shadow: none;
      }
    }

    .sendBtn {
      width: 20%;
      box-sizing: border-box;
      font-size: 1.25em;
      text-align: center;
      border: none;
      height: 100%;
      @include themify($themes) {
        background: themed('messageInputSendBtnBgColor');
        color: themed('messageInputSendBtnColor');
      }
      transition: all .35s ease-out;


    :disabled {
      opacity: .2;
      @include themify($themes) {
        background: themed('messageInputSendBtnDisabledBgColor');
      }
    }
    }
  }
}
/* END message input */
