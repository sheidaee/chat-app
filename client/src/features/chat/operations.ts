import { Dispatch } from 'redux';
import io from 'socket.io-client';

import {
  initSocket,
  userConnected,
  messageReceived,
  typing,
  addConnectedUsers,
  updateConnectedUsers,
  updateSettings,
  userReadMessages,
} from './actions';
import { SOCKET_URL } from '../../utilities/constants';
import { chatActionNameTypes, Chat, User, ConnectedUsers } from './types';

const startSocket = () => (dispatch: Dispatch) => {
  try {
    const socket = io(SOCKET_URL);

    socket.on('connect', () => {
      console.log('Connected');
    });

    dispatch(initSocket(socket));
  } catch (e) {
    console.log(e);
  }
};

interface VerifyUser {
  user: User;
  socket: any;
  callback: any;
}

const verifyUser = ({ user, socket, callback }: VerifyUser) => (
  dispatch: Dispatch
) => {
  try {
    socket.emit(chatActionNameTypes.VERIFY_USER, user, callback);
  } catch (e) {
    console.log(e);
  }
};

interface LoginUser {
  socket: any;
  user: User;
}

interface UserUpdated {
  users: ConnectedUsers;
  userId: string;
}

const loginUser = ({ socket, user }: LoginUser) => (dispatch: Dispatch) => {
  try {
    socket.emit(chatActionNameTypes.USER_CONNECTED, user);
    dispatch(userConnected(user));

    socket.on(
      chatActionNameTypes.USER_UPDATED,
      ({ users, userId }: UserUpdated) => {
        // debugger;
        dispatch(updateConnectedUsers({ users, userId }));
      }
    );
  } catch (e) {
    console.log(e);
  }
};

interface UpdateUser {
  socket: any;
  userId: string;
  newUserName: string;
  callback: any;
}

const updateUser = ({ socket, userId, newUserName, callback }: UpdateUser) => (
  dispatch: Dispatch
) => {
  try {
    socket.emit(
      chatActionNameTypes.UPDATE_USER,
      { userId, newUserName },
      callback
    );
  } catch (e) {
    console.log(e);
  }
};

interface UpdateTypingInChat {
  dispatch: Dispatch;
  chatId: string;
}

interface UpdateTypingInChatCallback {
  isTyping: boolean;
  userId: string;
}

const updateTypingInChat = ({ dispatch, chatId }: UpdateTypingInChat) => {
  return ({ isTyping, userId }: UpdateTypingInChatCallback) => {
    dispatch(typing({ isTyping, userId, chatId }));
  };
};

interface AddMessageToChat {
  dispatch: Dispatch;
  activeChat: string;
}

const addMessageToChat = ({ dispatch, activeChat }: AddMessageToChat) => {
  return (message: Chat) => {
    dispatch(messageReceived({ chat: message, reset: false, activeChat }));
    // dispatch(userReadMessages());
  };
};

interface AddChat {
  chat: Chat;
  reset: boolean;
  activeChat: string;
  dispatch: Dispatch;
  socket: any;
}

const addChat = ({ chat, reset, activeChat, dispatch, socket }: AddChat) => {
  dispatch(messageReceived({ chat, reset, activeChat }));

  const messageEvent = `${chatActionNameTypes.MESSAGE_RECEIVED}-${chat.id}`;
  const typingEvent = `${chatActionNameTypes.TYPING}-${chat.id}`;

  socket.on(typingEvent, updateTypingInChat({ dispatch, chatId: chat.id }));
  socket.on(messageEvent, addMessageToChat({ dispatch, activeChat }));
};

const resetChat = (dispatch: Dispatch, socket: any, chat: Chat) => {
  addChat({ chat, reset: true, activeChat: chat.name, dispatch, socket });
};

interface CommunityChat {
  socket: any;
}

const communityChat = ({ socket }: CommunityChat) => (dispatch: Dispatch) => {
  try {
    socket.emit(
      chatActionNameTypes.COMMUNITY_CHAT,
      resetChat.bind(null, dispatch, socket)
    );

    socket.on(chatActionNameTypes.USER_CONNECTED, (users: ConnectedUsers) => {
      dispatch(addConnectedUsers(users));
    });
  } catch (e) {
    console.log(e);
  }
};

interface SendMessage {
  message: string;
  socket: any;
  activeChatId: string;
}

const sendMessage = ({ message, socket, activeChatId }: SendMessage) => (
  dispatch: Dispatch
) => {
  try {
    socket.emit(chatActionNameTypes.MESSAGE_SENT, {
      chatId: activeChatId,
      message,
    });
  } catch (e) {
    console.log(e);
  }
};

interface SendTyping {
  chatId: string;
  socket: any;
  isTyping: boolean;
  formSubmitted?: boolean;
}

const sendTyping = ({
  chatId,
  socket,
  isTyping,
  formSubmitted,
}: SendTyping) => {
  try {
    // if (formSubmitted) {
    //   const typingEvent = `${chatActionNameTypes.TYPING}-${chatId}`;
    //   socket.off(typingEvent);
    //   return;
    // }

    socket.emit(chatActionNameTypes.TYPING, { chatId, isTyping });
  } catch (e) {
    console.log(e);
  }
};

export {
  startSocket,
  verifyUser,
  loginUser,
  sendMessage,
  communityChat,
  sendTyping,
  updateUser,
  updateSettings,
  userReadMessages,
};
