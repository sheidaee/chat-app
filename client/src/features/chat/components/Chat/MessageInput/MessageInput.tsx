import React, { useState, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Button } from '@blueprintjs/core';

import { chatOperations } from '../../..';

import TextAreaField from '../../../../../components/TextAreaField';
import useTranslation from '../../../../../hooks/useTranslation';
import './MessageInput.scss';

const MessageInput = () => {
  const dispatch = useDispatch();
  const socket = useSelector(({ app }: any) => app.socket);
  const chats = useSelector(({ app }: any) => app.chats);
  const activeChat = useSelector(({ app }: any) => app.activeChat);
  const sendOnCtrlEnterSettings = useSelector(
    ({ app }: any) => app.settings.sendMessageOnCtrlEnter
  );
  const { t } = useTranslation();

  const [message, setMessage] = useState('');
  const [isTyping, setIsTyping] = useState(false);

  const lastUpdateTime = useRef(-1);
  const typingIntervalRef = useRef<NodeJS.Timeout>();

  function stopCheckingTyping() {
    if (typingIntervalRef.current) {
      clearInterval(typingIntervalRef.current);
      chatOperations.sendTyping({
        isTyping: false,
        socket,
        chatId: chats[activeChat].id,
      });
    }
  }

  function startCheckingTyping() {
    typingIntervalRef.current = setInterval(() => {
      if (Date.now() - lastUpdateTime.current > 300) {
        setIsTyping(false);
        stopCheckingTyping();
      }
    }, 300);
  }

  async function sendTyping() {
    lastUpdateTime.current = Date.now();
    if (!isTyping) {
      setIsTyping(true);

      await chatOperations.sendTyping({
        isTyping: true,
        socket,
        chatId: chats[activeChat].id,
      });

      startCheckingTyping();
    }
  }

  const handleSubmit = (e: React.FormEvent<EventTarget>) => {
    e.preventDefault();

    dispatch(
      chatOperations.sendMessage({
        message,
        socket,
        activeChatId: chats[activeChat].id,
      })
    );

    setMessage('');
  };

  const handleKeyUp = async (e: React.KeyboardEvent) => {
    if (e.keyCode !== 13) {
      await sendTyping();
    }
  };

  const handleOnChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    setMessage(target.value);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>): void => {
    if (
      event.ctrlKey &&
      event.keyCode === 13 &&
      sendOnCtrlEnterSettings === '1'
    ) {
      event.preventDefault();
      event.stopPropagation();

      handleSubmit(event);
    }
  };

  return (
    <div className="messageInput">
      <form onSubmit={handleSubmit} className="messageForm">
        <TextAreaField
          name="message"
          id="message"
          className="formControl"
          value={message}
          autoComplete="off"
          placeholder={t('typePlaceHolder')}
          onKeyUp={handleKeyUp}
          onChange={handleOnChange}
          onKeyDown={handleKeyDown}
        />
        <Button disabled={message.length < 1} type="submit" className="sendBtn">
          {t('send')}
        </Button>
      </form>
    </div>
  );
};

export default MessageInput;
