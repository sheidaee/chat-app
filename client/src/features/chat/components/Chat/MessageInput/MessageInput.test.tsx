import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react';

import { renderWithRedux, generate, initSocket } from 'til-client-test-utils';
import MessageInput from './MessageInput';
import * as chatOperations from '../../../operations';

const newUser = generate.userData();
const user = { id: newUser.id, name: newUser.name };
const users = { [user.id]: user };

const socket = initSocket({});
jest.mock('../../../operations', () => {
  // const actualChatOperations = require.requireActual('../../../operations');

  return {
    sendMessage: jest.fn(() => {
      return jest.fn();
    }),
  };
});

beforeEach(jest.clearAllMocks);

afterEach(cleanup);

test('should send Messages', async () => {
  const messageItem = generate.messageData({ sender: user.id });

  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const { getByTestId, finishLoading, state } = renderWithRedux(
    <MessageInput />,
    {
      initialState,
    }
  );

  const messageNode = getByTestId('message');
  fireEvent.change(messageNode, { target: { value: messageItem.message } });
  getByTestId('sendMessageBtn').click();

  await finishLoading();

  const chat = state.app.chats[state.app.activeChat];
  const params = {
    activeChatId: chat.id,
    message: messageItem.message,
    socket: expect.any(Object),
  };

  expect(chatOperations.sendMessage).toBeCalledTimes(1);
  expect(chatOperations.sendMessage).toHaveBeenCalledWith(params);
});

test('should clean message field after sending Messages', async () => {
  const messageItem = generate.messageData({ sender: user.id });

  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const { getByTestId, finishLoading } = renderWithRedux(<MessageInput />, {
    initialState,
  });

  const messageNode = getByTestId('message');
  fireEvent.change(messageNode, { target: { value: messageItem.message } });
  getByTestId('sendMessageBtn').click();

  await finishLoading();

  expect(messageNode.value).toEqual('');
});

test('should not send Messages when on CTRL+ENTER is no allowed', async () => {
  const messageItem = generate.messageData({ sender: user.id });

  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1', sendMessageOnCtrlEnter: '1' },
  };

  const { getByTestId, finishLoading, state } = renderWithRedux(
    <MessageInput />,
    {
      initialState,
    }
  );

  const messageNode = getByTestId('message');
  fireEvent.change(messageNode, { target: { value: messageItem.message } });
  fireEvent.keyDown(messageNode, { ctrlKey: true, keyCode: 13 });

  await finishLoading();

  const chat = state.app.chats[state.app.activeChat];
  const params = {
    activeChatId: chat.id,
    message: messageItem.message,
    socket: expect.any(Object),
  };

  expect(chatOperations.sendMessage).toBeCalledTimes(1);
  expect(chatOperations.sendMessage).toHaveBeenCalledWith(params);
});

test('should send Messages on CTRL+ENTER', async () => {
  const messageItem = generate.messageData({ sender: user.id });

  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1', sendMessageOnCtrlEnter: '2' },
  };

  const { getByTestId, finishLoading, state } = renderWithRedux(
    <MessageInput />,
    {
      initialState,
    }
  );

  const messageNode = getByTestId('message');
  fireEvent.change(messageNode, { target: { value: messageItem.message } });
  fireEvent.keyDown(messageNode, { ctrlKey: true, keyCode: 13 });

  await finishLoading();

  expect(chatOperations.sendMessage).toBeCalledTimes(0);
});

test('should not send Messages when there is no message', async () => {
  const { getByTestId, finishLoading, container } = renderWithRedux(
    <MessageInput />
  );

  const messageNode = getByTestId('message');
  fireEvent.change(messageNode, { target: { value: '' } });
  getByTestId('sendMessageBtn').click();

  await finishLoading();

  expect(chatOperations.sendMessage).toBeCalledTimes(0);
});

test('snapshot', async () => {
  const chats: any = {
    community: {
      id: generate.id(),
      messages: [],
      typingUsers: [],
    },
  };

  const initialState = {
    socket,
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const { container } = renderWithRedux(<MessageInput />, {
    initialState,
  });

  expect(container).toMatchSnapshot();
});
