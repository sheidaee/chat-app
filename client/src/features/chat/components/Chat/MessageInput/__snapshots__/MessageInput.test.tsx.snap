// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    class="messageInput"
  >
    <form
      class="messageForm"
      data-testid="messageForm"
    >
      <textarea
        autocomplete="off"
        class="bp3-input formControl"
        data-testid="message"
        id="message"
        name="message"
        placeholder="Enter message"
      />
      <button
        class="bp3-button bp3-disabled sendBtn"
        data-testid="sendMessageBtn"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        <span
          class="bp3-button-text"
        >
          Send
        </span>
      </button>
    </form>
  </div>
</div>
`;
